### PR TITLE
Mldb 2120 procedure deadlock test

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: arch/rcu_protected.h [hunk 1]
```diff
@@ -199,8 +199,9 @@ struct RcuProtected {
         return std::unique_ptr<T>(val.exchange(newVal));
     }
     
+    template<typename DELETER>
     bool cmp_xchg(RcuLocked<T> & current,
-                  std::unique_ptr<T> & newValue,
+                  std::unique_ptr<T, DELETER> & newValue,
```
Comment (FinchPowers) [inline arch/rcu_protected.h]: Not related to this PR, but adds flexibility. (And I used it while debugging.)

File: rest/rest_collection_impl.h [hunk 2]
```diff
@@ -561,7 +561,8 @@ addBackgroundJobInThread(Key key,
 
         // Set up the task, without starting it yet
         auto task = std::make_shared<BackgroundTask>();
-        task->config = std::move(config);
+        task->running = false;
+        task->config = config;
```
Comment (FinchPowers) [inline rest/rest_collection_impl.h]: std::move(config)... if it did not work the first time, the config was lost.

File: rest/rest_collection_impl.h [hunk 3]
```diff
@@ -615,6 +616,7 @@ addBackgroundJobInThread(Key key,
 
         if (impl->entries.cmp_xchg(oldEntries, newEntries, true)) {
             // Now we can start the task, since the commit succeeded
+            task->running = true;
```
Comment (FinchPowers) [inline rest/rest_collection_impl.h]: As long as it doesn't enter this condition scope, the task was not really running, hence the deadlock.

File: rest/rest_collection_impl.h (hunk 2 above)
Comment (guyd) [inline rest/rest_collection_impl.h]: Good to know.

**Comment by jeremybarnes** [comment]
Nice work Mich!

Please consider:
1.  Making the destructor noexcept, by catching an exception from cancel() and aborting immediately.  In the more distant future, cancel() should be made noexcept, but that will require deeper changes.
2.  Adding a test to rest_collection_test to reproduce the condition without the baggage of the rest of MLDB.  There are already other stress tests there.

File: rest/rest_collection_impl.h [hunk 4]
```diff
@@ -561,7 +561,8 @@ addBackgroundJobInThread(Key key,
 
         // Set up the task, without starting it yet
         auto task = std::make_shared<BackgroundTask>();
-        task->config = std::move(config);
+        task->running = false;
```
Comment (guyd) [inline rest/rest_collection_impl.h]: Shouldn't we set it to false at construction?  I am surprised that this was not the case.

Comment (FinchPowers) [inline rest/rest_collection_impl.h]: Possibly. It crossed my mind but I did not ask. I'll dig it.

**Comment by FinchPowers** [comment]
@jeremybarnes So far I was unsuccessful at reproducing the issue without the MLDB layer. I had to use inception which was both slow and time consuming enough to cause the issue.

I am opened to advice on how to try to reproduce it without the MLDB layer.

**Pull request merged**